File: ios/MuscleCheck/viewModels/HistoryViewModel.swift
//
//  HistoryViewModel.swift
//  MuscleCheck
//
//  Created by Alejandro De Jesus on 07/06/2025.
//


import Foundation
import SwiftData

final class HistoryViewModel: ObservableObject {

    /// The day whose week is highlighted and whose breakdown is shown below the calendar.
    @Published var selectedDate: Date = Date()
    /// The month currently rendered in the grid (changed by the chevrons).
    @Published var displayedMonth: Date = Date()
    /// Calendar starts collapsed to the selected week; tapping the header expands it to the full month.
    @Published var isCalendarExpanded: Bool = false
    @Published var entries: [MuscleEntry]

    init(entries: [MuscleEntry]) {
        self.entries = entries
    }

    static func create(with entries: [MuscleEntry]) -> HistoryViewModel {
        return HistoryViewModel(entries: entries)
    }

    // MARK: - Calendar grid (delegates to MonthCalendarCalculator)

    var weeks: [[CalendarDay]] {
        MonthCalendarCalculator.monthMatrix(for: displayedMonth)
    }

    var intensityByDay: [Date: Int] {
        MonthCalendarCalculator.muscleCountByDay(from: entries)
    }

    var monthTrainedCount: Int {
        MonthCalendarCalculator.trainedDayCount(inMonthOf: displayedMonth, from: entries)
    }

    var weekBreakdown: [DayActivities] {
        MonthCalendarCalculator.weekBreakdown(forWeekContaining: selectedDate, from: entries)
    }

    /// The single Monday-first week shown when the calendar is collapsed.
    var selectedWeek: [CalendarDay] {
        MonthCalendarCalculator.weekRow(forWeekContaining: selectedDate)
    }

    /// "Junio 2026" — capitalized for the header.
    var monthTitle: String {
        Self.monthTitleFormatter.string(from: displayedMonth).capitalized
    }

    /// "junio" — lowercase, used mid-sentence in the month summary caption.
    var monthName: String {
        Self.monthNameFormatter.string(from: displayedMonth)
    }

    // MARK: - Navigation

    func goToPreviousMonth() {
        if let prev = Date.appCalendar.date(byAdding: .month, value: -1, to: displayedMonth) {
            displayedMonth = prev
        }
    }

    func goToNextMonth() {
        if let next = Date.appCalendar.date(byAdding: .month, value: 1, to: displayedMonth) {
            displayedMonth = next
        }
    }

    func select(_ day: Date) {
        selectedDate = day
        // Tapping a leading/trailing day follows it into its month (Calendar.app behavior).
        if !Date.appCalendar.isDate(day, equalTo: displayedMonth, toGranularity: .month) {
            displayedMonth = day
        }
        // Picking a day collapses back to its week (Calendar.app behavior).
        isCalendarExpanded = false
    }

    func goToPreviousWeek() { shiftSelectedWeek(by: -1) }
    func goToNextWeek() { shiftSelectedWeek(by: 1) }

    /// Moves the collapsed view a week at a time, keeping `displayedMonth` (and the header
    /// label) in step with the visible week.
    private func shiftSelectedWeek(by value: Int) {
        guard let next = Date.appCalendar.date(byAdding: .weekOfYear, value: value, to: selectedDate) else { return }
        selectedDate = next
        displayedMonth = next
    }

    func toggleCalendarExpanded() {
        isCalendarExpanded.toggle()
        // Expanding shows the month around the selected day.
        if isCalendarExpanded { displayedMonth = selectedDate }
    }

    // MARK: - Formatters

    private static let monthTitleFormatter: DateFormatter = {
        let f = DateFormatter()
        f.setLocalizedDateFormatFromTemplate("LLLL yyyy")
        return f
    }()

    private static let monthNameFormatter: DateFormatter = {
        let f = DateFormatter()
        f.setLocalizedDateFormatFromTemplate("LLLL")
        return f
    }()
}

File: android/app/src/main/java/com/zadkiel/musclecheck/ui/history/HistoryViewModel.kt
package com.zadkiel.musclecheck.ui.history

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.zadkiel.musclecheck.data.prefs.UserPreferencesRepository
import com.zadkiel.musclecheck.data.repository.MuscleRepository
import com.zadkiel.musclecheck.domain.MonthCalendarCalculator
import com.zadkiel.musclecheck.domain.model.CustomCategory
import com.zadkiel.musclecheck.domain.model.MuscleEntry
import com.zadkiel.musclecheck.domain.model.WeightUnit
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.SharingStarted
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.combine
import kotlinx.coroutines.flow.stateIn
import java.time.LocalDate
import java.time.YearMonth
import java.time.format.DateTimeFormatter
import java.util.Locale
import javax.inject.Inject

data class HistoryUiState(
    val entries: List<MuscleEntry> = emptyList(),
    val customCategories: List<CustomCategory> = emptyList(),
    val weightUnit: WeightUnit = WeightUnit.KG,
    val selectedDate: LocalDate = LocalDate.now(),
    val displayedMonth: LocalDate = LocalDate.now(),
    val isCalendarExpanded: Boolean = false,
) {
    val weeks get() = MonthCalendarCalculator.monthMatrix(displayedMonth)
    val selectedWeek get() = MonthCalendarCalculator.weekRow(selectedDate)
    val intensityByDay get() = MonthCalendarCalculator.muscleCountByDay(entries)
    val monthTrainedCount get() = MonthCalendarCalculator.trainedDayCount(displayedMonth, entries)
    val weekBreakdown get() = MonthCalendarCalculator.weekBreakdown(selectedDate, entries)

    /** "Junio 2026" — capitalized for the header. */
    val monthTitle: String
        get() = DateTimeFormatter.ofPattern("LLLL yyyy", Locale.getDefault())
            .format(displayedMonth)
            .replaceFirstChar { if (it.isLowerCase()) it.titlecase(Locale.getDefault()) else it.toString() }

    /** "junio" — lowercase, used mid-sentence in the month summary caption. */
    val monthName: String
        get() = DateTimeFormatter.ofPattern("LLLL", Locale.getDefault()).format(displayedMonth)
}

@HiltViewModel
class HistoryViewModel @Inject constructor(
    repository: MuscleRepository,
    prefs: UserPreferencesRepository,
) : ViewModel() {

    private val selection = MutableStateFlow(
        Triple(LocalDate.now(), LocalDate.now(), false) // selectedDate, displayedMonth, isExpanded
    )

    val uiState: StateFlow<HistoryUiState> = combine(
        repository.entries,
        repository.customCategories,
        prefs.weightUnit,
        selection,
    ) { entries, custom, unit, (selected, month, expanded) ->
        HistoryUiState(
            entries = entries,
            customCategories = custom,
            weightUnit = unit,
            selectedDate = selected,
            displayedMonth = month,
            isCalendarExpanded = expanded,
        )
    }.stateIn(viewModelScope, SharingStarted.WhileSubscribed(5_000), HistoryUiState())

    fun select(day: LocalDate) {
        val (_, month, _) = selection.value
        // Tapping a leading/trailing day follows it into its month; picking a day
        // collapses back to its week (Calendar.app behavior).
        val newMonth = if (YearMonth.from(day) != YearMonth.from(month)) day else month
        selection.value = Triple(day, newMonth, false)
    }

    fun goToPreviousMonth() = shiftMonth(-1)
    fun goToNextMonth() = shiftMonth(1)

    private fun shiftMonth(value: Long) {
        val (selected, month, expanded) = selection.value
        selection.value = Triple(selected, month.plusMonths(value), expanded)
    }

    fun goToPreviousWeek() = shiftSelectedWeek(-1)
    fun goToNextWeek() = shiftSelectedWeek(1)

    /** Moves the collapsed view a week at a time, keeping the header label in step. */
    private fun shiftSelectedWeek(value: Long) {
        val (selected, _, expanded) = selection.value
        val next = selected.plusWeeks(value)
        selection.value = Triple(next, next, expanded)
    }

    fun toggleCalendarExpanded() {
        val (selected, month, expanded) = selection.value
        // Expanding shows the month around the selected day.
        selection.value = Triple(selected, if (!expanded) selected else month, !expanded)
    }
}
